src/node_api.h

```diff
@@ -44,12 +44,34 @@ typedef struct napi_module {
 #define NAPI_MODULE_VERSION 1
 
 #if defined(_MSC_VER)
+#if defined(__cplusplus) && defined(__cpp_inline_variables)
+// The NAPI_C_CTOR macro defines a function fn that is called during dynamic
+// initialization of static variables.
+// The order of the dynamic initialization is not defined and code in fn
+// function must avoid using other static variables with dynamic initialization.
+#define NAPI_C_CTOR(fn)                                                        \
+  static void __cdecl fn(void);                                                \
+  namespace {                                                                  \
+  struct fn##_ {                                                               \
+    static int Call##fn() { return (fn(), 0); }                                \
+    static inline const int x = Call##fn();                                    \
```

Comment:
@Flarna, no, we do not need it in this context. It is only required for the changes in the node.h to support the shared mode. Since the support for VS2015 is required for addons, I am going to refactor the code in a way that only that shared mode is going to require the C++17 and all other places are going to use the method you proposed above,

Reply:
Why is it needed in node.h but not here? What's the difference between using napi or node in this regard?

Reply:
node.h has macro `NODE_SHARED_MODE` that removes the `static` prefix for the registration function.
I assume this is done to support using the `NODE_C_CTOR` macro in header files.
With functions that approach works, but initialization of static class variables which we want to use in MSVC C++ it does not work. Thus, I decided to use the new C++17 `inline static` that was created for such scenarios. My hope was that C++17 is the requirement that we have for Node.js code and its modules.
I am going to research it a little bit more: I believe the C++11 static field initialization in headers can be addressed by using templates. If not, then I will keep use of C++17 `inline static` only for `NODE_SHARED_MODE` and use simple static field initialization for other scenarios.

Reply:
It seems that the shared mode introduced in https://github.com/nodejs/node/commit/410296c37acb2ad5d939bd4633b2c3fbe3782084 was required for compiling Node.js as a shared module. It has nothing to do with being able to use registration in header files. Thus, no complexity needed.